### ISSUE TITLE
Fix required for the GDPR Vulnerabilities detected by Appscan

### ISSUE THREAD

**Issue by Vyshakh18** [mozilla/rhino#425]
As a part of **GDPR** compliance exercise we recently scanned the Mozilla Rhino Javascript jar and found that **51 vulnerabilities** were detected by the **Appscan** software.
Kindly requesting to address these issues.
I am attaching the zipped scan report which contains the details of the vulnerabilities
[Rhino 1.7.9.zip](https://github.com/mozilla/rhino/files/1928112/Rhino.1.7.9.zip)

**Comment by sainaen** [comment]
Thanks for reporting!

To make this a bit easier to view, I published the contents of your zip file [on Github Pages.](https://sainaen.github.io/rhino-appscan-results/) (hope that’s ok with you)

So, there are 41 _PrivilegeEscalation_-s that point at lines with method being called through reflection (i.e. `method.invoke(…)`). Do you know or have any pointers on how to perform a privilege escalation with that?

The other 10 issues are all within `org.mozilla.javascript.tools` package and I think these are false positives. For example, there’s _PathTraversal_ finding in the implementation of [`readUrl()` method](https://github.com/mozilla/rhino/blob/master/toolsrc/org/mozilla/javascript/tools/shell/Global.java#L1099) available in shell, but this method is _designed_ to read arbitrary paths.

**Comment by Vyshakh18** [comment]
Hi,
Thanks for your quick response.
The only information i have in this regard is this link:
http://cwe.mitre.org/data/definitions/266.html
I am not sure if this would help.

**Comment by gbrail** [comment]
Similarly, the tool is complaining about writing to file in code that's part of a tool that's used during development to write files (IdSwitch). 

As for the "PrivilegeEscalation" -- that seems to flag every time Rhino uses reflection to invoke a method. Given the nature of Rhino -- it generates bytecode that it then invokes via reflection -- it's hard to imagine a version of Rhino that would work without invoking methods that way.

Does anyone in the community have any other context on why such a tool would flag this and what one might do to handle this differently?

Finally I am not a lawyer but I am not aware of what any of this has to do with GDPR.

**Comment by sainaen** [comment]
I have a theory: it basically complains about all the ways it thinks is possible to _trick_ Rhino into calling methods from within `org.mozilla.javascript` package, which could potentially circumvent the `SecurityManager` that would not allow such invocations directly. For example via `ScriptableObject`’s public `defineClass()` we could call `public static void init(Scriptable scope)` of a given class.

---

(By the way, according to git log, 19th of April is an open-source birthday of Mozilla Rhino. :tada: Though, confusingly, the sources differ on the year — it’s either [1998](https://developer.mozilla.org/en-US/docs/Mozilla/Projects/Rhino/History) or [1999](https://github.com/mozilla/rhino/commit/df654e87).)

**Comment by gbrail** [comment]
Yep -- I'm sure that's what it's related to. If you are running Rhino in a
multi-tenant JVM or one in which the Java security manager is present, all
those reflections could theoretically result in an escalation of privilege.
I'm not sure, however, if there is a bug that we can "fix.."

**Comment by Vyshakh18** [comment]
@gbrail - so do you mean, this will be looked upon further and there is a possibility to get fix(es) for these vulnerabilities?

**Comment by gbrail** [comment]
Here's a summary of my analysis.

The Injection.OS, PathTraversal, InsecureAlgorithm, and CrossSiteScripting vulnerabilities are all in code that is only used by the Rhino interactive shell, the "Global" class, and the command-line tools like jsc and idswitch. None of these things should be used in a sensitive environment. (And they are typically not used.)

The tool appears to flag all uses of Java reflection as possible PrivilegeEsclation vulnerabilities. Reflection in Java can be used to escalate privilege if the calling code can get access to a private or protected field or method and make it possible to invoke that method or access that field via a public method. 

I inspected all the use of reflection in Rhino and found that in all cases Rhino is only using these mechanism to invoke public methods. (For instance, the code in ScriptableObject that generates a JavaScript wrapper around a Java class looks only for public methods of the class, so it cannot be used to invoke private methods or access private fields.)

I can suggest one specific fix that we can make: We can resurrect the so called "small JAR" distribution of Rhino. This results in a JAR that does not include the tools. Such a JAR would remove all the code that has all the vulnerabilities save for "PrivilegeEscalation". If that would help, it's not a difficult thing to do.

Otherwise, I would be quite happy for someone in the community to do a similar analysis (or run a similar tool) and either report specific things that can be fixed, or submit PRs to fix them.

Thanks!

**Comment by Vyshakh18** [comment]
Thanks a lot @gbrail 
Could you suggest how to build the 'small jar' ?
